pkg/document/crdt/element_rht.go

```diff
@@ -0,0 +1,211 @@
+// to the outside.
+type ElementRHT struct {
+	nodeMapByKey       map[string]*ElementRHTNode
+	nodeMapByCreatedAt map[string]*ElementRHTNode
+}
+
+// NewElementRHT creates a new instance of ElementRHT.
+func NewElementRHT() *ElementRHT {
+	return &ElementRHT{
+		nodeMapByKey:       make(map[string]*ElementRHTNode),
+		nodeMapByCreatedAt: make(map[string]*ElementRHTNode),
+	}
+}
+
+// Get returns the value of the given key.
+func (rht *ElementRHT) Get(key string) Element {
+	if node, ok := rht.nodeMapByKey[key]; ok {
+		if node.isRemoved() {
+			return nil
+		}
+		return node.elem
+	}
+	return nil
+}
+
+// Has returns whether the element exists of the given key or not.
+func (rht *ElementRHT) Has(key string) bool {
+	if node, ok := rht.nodeMapByKey[key]; ok {
+		return node != nil && !node.isRemoved()
+	}
+	return false
+}
+
+// Set sets the value of the given key. If there is an existing value, it is removed.
+func (rht *ElementRHT) Set(k string, v Element) Element {
+	node, ok := rht.nodeMapByKey[k]
+	var removed Element
+	if ok && !node.isRemoved() && v.CreatedAt().After(node.elem.CreatedAt()) {
+		node.Remove(v.CreatedAt())
+		removed = node.elem
+	}
+	newNode := newElementRHTNode(k, v)
+	rht.nodeMapByCreatedAt[v.CreatedAt().Key()] = newNode
+	if !ok || v.CreatedAt().After(node.elem.CreatedAt()) {
+		rht.nodeMapByKey[k] = newNode
+	}
+
+	return removed
+}
+
+// Delete deletes the Element of the given key.
+func (rht *ElementRHT) Delete(k string, deletedAt *time.Ticket) Element {
+	node, ok := rht.nodeMapByKey[k]
+	if !ok {
+		return nil
+	}
+
+	if !node.Remove(deletedAt) {
+		return nil
+	}
+
+	return node.elem
+}
+
+// DeleteByCreatedAt deletes the Element of the given creation time.
+func (rht *ElementRHT) DeleteByCreatedAt(createdAt *time.Ticket, deletedAt *time.Ticket) Element {
+	node, ok := rht.nodeMapByCreatedAt[createdAt.Key()]
+	if !ok || node == nil {
+		return nil
+	}
+
+	if !node.Remove(deletedAt) {
+		return nil
+	}
+
+	return node.elem
+}
+
+// Elements returns a map of elements because the map easy to use for loop.
+// TODO: If we encounter performance issues, we need to replace this with other solution.
+func (rht *ElementRHT) Elements() map[string]Element {
+	members := make(map[string]Element)
+	for _, node := range rht.nodeMapByKey {
+		if !node.isRemoved() {
+			members[node.key] = node.elem
+		}
+	}
+
+	return members
+}
+
+// Nodes returns a map of elements because the map easy to use for loop.
+// TODO: If we encounter performance issues, we need to replace this with other solution.
+func (rht *ElementRHT) Nodes() []*ElementRHTNode {
+	var nodes []*ElementRHTNode
+	for _, node := range rht.nodeMapByKey {
+		nodes = append(nodes, node)
+	}
+
+	return nodes
+}
+
+// purge physically purge child element.
+func (rht *ElementRHT) purge(elem Element) {
+	node, ok := rht.nodeMapByCreatedAt[elem.CreatedAt().Key()]
+	if !ok {
+		panic("fail to find: " + elem.CreatedAt().Key())
+	}
+	if node == nil {
+		return
+	}
```

Comment:
We are checking existence with `ok`, but is there a reason to separately check if `node` is nil?

Reply:
I had a nil check to handle exceptions while I was writing this pr, but I don't think we need it(And other commented things too) now.